src/main/java/backend/Logic.java

```diff
@@ -264,9 +265,20 @@ public Model getRepo(String repoId) {
                 repoOpControl.replaceIssueLabelsLocally(issue, newLabels);
         localLabelsReplaceFuture.thenRun(this::refreshUI);
 
-        return updateIssueLabelsOnServer(issue, newLabels)
+        updateIssueLabelsOnServer(issue, newLabels)
+                .exceptionally((e) -> {
+                    result.completeExceptionally(e);
+                    return false;
+                })
                 .thenCombine(localLabelsReplaceFuture, (isUpdateSuccessful, locallyModifiedIssue) ->
-                        handleIssueLabelsUpdateResult(isUpdateSuccessful, locallyModifiedIssue, originalLabels));
+                        handleIssueLabelsUpdateResult(isUpdateSuccessful, locallyModifiedIssue, originalLabels))
+                .thenAccept((isReplacementSuccessful) -> {
+                    if (!result.isCompletedExceptionally()) {
```

Comment:
There are two `Futures` here because we want to handle the exception both in `Logic` (reverting to the previous labels) and then still propagate it to the `UI` (`exceptionally` will end the exception propagation). That's the also the reason why the `exceptionally` is put in the middle. I will see how this can be made clearer.